halo2_proofs/src/plonk/evaluation.rs

```diff
@@ -185,7 +187,7 @@ pub struct Evaluator<C: CurveAffine> {
     ///  Custom gates evalution
     pub custom_gates: GraphEvaluator<C>,
     ///  Lookups evalution
-    pub lookups: Vec<GraphEvaluator<C>>,
+    pub lookups: Vec<(Vec<GraphEvaluator<C>>, GraphEvaluator<C>)>,
```

Comment:
Could you explain me the purpose of the "GraphEvaluator" data type (the line 200, if you already know the necessary details)? Perhaps, I will have to deal with it, though still not sure.

Reply:
It's a helper struct to do all the necessary evaluations in a more optimal way. So our circuits have lots of expressions, and most of them often overlap. So this tries to find sub expressions that are shared between all these expressions so it can optimize the calculation of all necessary expressions. Simple example: we have two expressions:
- `a + b + c`
- `a + b + d`

In a naive way, this requires 4 additions (two additions for each expression). However `a + b` is shared between the two expressions. So what the `GraphEvaluator` tries to do is calculate `t := a + b` once, and then calculate `t + c` and `t + d` to get the result of both expressions, but now only using 3 additions.